File: spring-data-requery/src/test/java/org/springframework/data/requery/domain/sample/AbstractChild.java
package org.springframework.data.requery.domain.sample;

import io.requery.Entity;
import io.requery.Generated;
import io.requery.Key;
import io.requery.ManyToMany;
import lombok.Getter;
import org.springframework.data.requery.domain.AbstractPersistable;

import java.util.Set;

/**
 * AbstractChild
 *
 * @author debop@coupang.com
 * @since 18. 6. 25
 */
@Getter
@Entity
public abstract class AbstractChild extends AbstractPersistable<Long> {

    @Key
    @Generated
    protected Long id;

    @ManyToMany(mappedBy = "children")
    protected Set<Parent> parents;

    protected String name;

    public void addParent(Parent parent) {
        this.getParents().add(parent);
    }

    private static final long serialVersionUID = 4307155877990688374L;
}

File: spring-data-requery-kotlin/src/test/kotlin/com/coupang/springframework/data/requery/domain/sample/AbstractChild.kt
package com.coupang.springframework.data.requery.domain.sample

import com.coupang.springframework.data.requery.domain.AbstractPersistable
import io.requery.Entity
import io.requery.Generated
import io.requery.Key
import io.requery.ManyToMany

/**
 * com.coupang.springframework.data.requery.domain.sample.AbstractChild
 * @author debop
 * @since 18. 5. 30
 */
@Entity
abstract class AbstractChild: AbstractPersistable<Long>() {

    @get:Key
    @get:Generated
    abstract val id: Long?

    @get:ManyToMany(mappedBy = "children")
    abstract val parents: MutableSet<AbstractParent>

    fun addParent(parent: AbstractParent): AbstractChild {
        this.parents += parent
        if(!parent.children.contains(this)) {
            parent.addChild(this)
        }
        return this
    }
}